### ISSUE TITLE
Switch Bluesky to opt out

### ISSUE THREAD

**Issue by snarfed** [snarfed/bridgy-fed#1471]
Not actually planned yet, but lots of people regularly request it, so I want a place to capture some of the discussion and ideas.

Norms on Bluesky are somewhat different from the fediverse in this regard. Also, unlike the fediverse, Bluesky is a fully public network, with global full text search, and the whole network's data is easily downloadable and accessible. And finally, the team themselves have said that they want it to be an open network, and prefer tools and bridges like BF to be opt out.

On the other hand, https://snarfed.org/2024-11-01_53932 . I wouldn't really consider making this change until if/when Bridgy Fed has some kind of real organizational structure, governance, and sustainability, as opposed to just one person's side project like it is now.

**Comment by snarfed** [comment]
If this ever happens, we'd need an _extremely_ easy way for Bluesky users to opt out. We already interpret [_Discourage apps from showing my account to logged-out users_](https://bsky.app/moderation) as opting out, but we'd probably also want a way that's Bridgy Fed specific, trivial to do, and minimally invasive (ie not a hashtag in profile bio.) DMing _no_ or _opt out_ to [@ap.brid.gy](https://bsky.app/profile/ap.brid.gy) is one option. OAuthed button on https://fed.brid.gy/ could be another.

**Comment by devilspeech** [comment]
This would likely alleviate a lot of the friction that opt in has introduced. My experience of trying to ask people to opt in on Bluesky has been like pulling teeth, even if everyone has been fine with it once they know it's an option. Most Bluesky users don't have DMs open since that is the default setting (as opposed to fedi where most people don't bother to turn on the setting that blocks DM notifications) so I am essentially forced to have an account on Bluesky to inform people about the bridge. And even then people often don't seem to understand what to do despite my best efforts to explain it in simple terms... Not to mention language barriers, on the off chance that someone does have DMs open the DM that Bridgy sends is also only in English.

I know it has been discussed ad nauseam, but I hope this could be reevaluated with more perspectives from people who are actually trying to use the bridge now and are running up against these barriers. I have accepted that we just have to live with it on fedi, but it would help us a great deal too if Bluesky didn't have the same artificial restrictions imposed on them.

**Comment by snarfed** [comment]
Thanks @devilspeech!

One difficulty here is that people often complain and provide feedback when they're unhappy more than when they're happy. I may consistently hear from people right now who want Bridgy Fed to be opt out, but since it's opt in, I may not get a proportionate amount of feedback from people who want it to stay opt in.

Not that volume of feedback for or against is the main deciding factor here; just trying to be aware that it may not be representative. Hard to know.

**Comment by shiribailem** [comment]
Just going to chime in that the norms aren't different, you just got dogpiled by random people who don't know the norms. Typical vocal minority sort of thing because a lot of people who recently joined think the fediverse is some sort of closed secret thing...

**Comment by devilspeech** [comment]
Exactly, it was clear from the start that the fedi users who demanded to be "opted out" simply didn't understand or were being willfully obtuse about how federation works. The bridge is like any other instance, no data is sent unless users choose to follow each other, and if you don't want to federate you can block the domain at the single click of button. Most of the instances and users who would object to being bridged probably already have the domain blocked anyway, so what functional purpose does opt in even serve for them now? They aren't the ones who have to jump through all the hoops they wanted put in place. All their harassment campaign really accomplished was making it so the rest of the millions of people on these networks who are desperate to maintain their connections are functionally barred from doing so. All because a minority of fedi users were confused and willing to resort to bullying to get what they wanted.

Again, I understand that it's no longer up for debate and at this point it's more about the fact that Bridgy is not critical infrastructure that it has to stay this way. It's just frustrating knowing that it is entirely feasible for Bluesky and fedi to interoperate and if things had only gone slightly differently they would be connected by now. It feels even more important at this point in time when so many people are in the process of trying to move off of other platforms and need all the encouragement they can get.

**Comment by Fauli1221** [comment]
I personally would recommend to post about this issue on bluesky and try to get as many opinions as possible 
if there is a big pushback it's a bad idea if there is just a few people who don't want it and most people don't mind or care you can consider it

**Comment by Tamschi** [comment]
@shiribailem @devilspeech It's not just a minority. You can see this for example by comparing with Flipboard who got mass-defederated recently over bot activity. For technical reasons, Bridgy Fed does have to run a very similar follow-bot to provide a consistent bridging experience from ActivityPub. The main difference is that this one only follows (back) when prompted, which makes it more acceptable to admins.

That said, have a look at #1305. An admin opting their instance in by default is much more technically feasible and shouldn't lead to backlash if it's communicated well, while still removing most of the friction. (In fact Flipboard is doing that already, just with a custom implementation to simulate individual opt-ins towards Bridgy Fed.)
Doing it via the ActivityPub Relay mechanism also would mean much lower server load and that we wouldn't need the follow bot to bridge users from that instance consistently (outside of possibly 'Quiet public' replies to bridged posts (see #1036), but it could be better to still use individual opt-in for those anyway).

Before bridging groups of accounts without individual opt-in though (in either direction), I think it's important to have #1406 first.
Bluesky does have pretty decent safety tooling now, arguably better than Mastodon, but to use it the bridge accounts need to be able to block anyone on the network.

**Comment by devilspeech** [comment]
For bridging AT to AP I had assumed that an opt out system would work similarly to bird.makeup where the AP account for the user would only be created when someone searched for that user on their instance, so it would only bridge Bluesky accounts that were followed by real people on the fediverse. I have to imagine the bot wouldn't need to follow every single user on Bluesky in order to have opt out for Bluesky? If being followed by the bot is actually required for fedi users to be searchable on Bluesky at all that's fair, opt in for fedi makes sense then, but if it's possible to have a system where we can just straightforwardly search for and follow Bluesky users on fedi without having to contact every single individual person to get them to opt in to the bridge that would be ideal, in my opinion. Fedi users are typically pretty good at figuring out how to use these kinds of tools if they really want to anyway.

**Comment by snarfed** [comment]
> For bridging AT to AP I had assumed that an opt out system would work similarly to bird.makeup where the AP account for the user would only be created when someone searched for that user on their instance, so it would only bridge Bluesky accounts that were followed by real people on the fediverse. I have to imagine the bot wouldn't need to follow every single user on Bluesky in order to have opt out for Bluesky?

That's right. @Tamschi was talking about the other direction, bridging from fediverse to Bluesky, and how fediverse  admins could currently automatically bridge everyone on their instance, ie #1305 .

**Comment by Tamschi** [comment]
@snarfed was a little faster, but yes that's all accurate.

On fedi/ActivityPub, searching for a Bluesky user with their full bridge handle or id will "probe" the bridge. We could even return the profile without actually enabling the bridge for that user yet, and do so only when someone follows. It's not necessary to follow someone to observe their activity on ATProto.
(I wonder if we can see who performs the lookup, but I suspect in many cases the profile fetch will be signed by the instance actor instead.)

On Bluesky, searches are completely invisible to Bridgy Fed. The account usually has to be actively 'pushed' into the ATProto network before it becomes visible there. We also can't see fediverse accounts or activity _at all_ without at least one direct follow or an AP Relay connection from that instance.
(Some can be discovered because it's mentioned in other activities, so it _is_ often possible to crawl ActivityPub, but that's not enough to bridge anything in more-or-less real time.)

**Comment by devilspeech** [comment]
I'm curious then, why exactly was Bluesky also made opt in from the start if all of the technical limitation is on fedi's side and there seemingly wasn't much if any personal objection from Bluesky's end? Was there even that much backlash specifically over the prospect of Bluesky users being bridged into fedi? Even then if an admin or user objects to that it's simple enough to just block the bridge domain and never have to see them, that's just normal inter-instance fedi moderation at the end of the day.

**Comment by Tamschi** [comment]
Something that's also relevant here is #971, but that _might_ be within reason for instance moderators to handle by silencing accounts that don't self-label (or the `bsky.brid.gy` domain entirely, more realistically) from the federated feed.

Even if #971 is implemented, it's not against the rules to post without self-labelling adult or graphic content on Bluesky, so I assume most admins would limit the instance at least a little bit to deal with posts that don't have a CW when they first arrive.

**Comment by shiribailem** [comment]
> @shiribailem @devilspeech It's not just a minority. You can see this for example by comparing with Flipboard who got mass-defederated recently over bot activity.

Where are we getting this from? This like the "mass" defederation of Threads (a minority of admins throwing a fit, but at least they had some reason)?

Public opt-out bridging has been a part of federation from the beginning a part of the culture... the people complaining about the idea of it are the same kind of people who call the entire ActivityPub Fediverse "Mastodon".

**Comment by devilspeech** [comment]
> Something that's also relevant here is #971, but that _might_ be within reason for instance moderators to handle by silencing accounts that don't self-label (or the `bsky.brid.gy` domain entirely, more realistically) from the federated feed.
> 
> Even if #971 is implemented, it's not against the rules to post without self-labelling adult or graphic content on Bluesky, so I assume most admins would limit the instance at least a little bit to deal with posts that don't have a CW when they first arrive.

Yeah I have encountered this before, although so far I've seen more people who have self-labeled than those who haven't. It's still to be seen if Bluesky will develop a culture of consistently self-labeling or not. It's also an option for admins to apply sensitive media flags to individual posts (if reported) or to entire accounts so all of the media they upload is flagged. This is how my instance handles it for now if anyone reports Bridgy posts that don't have CWs.

This is something fedi already has to deal with regardless of whether the bridge is opt in or opt out though. Silencing is a reasonable choice for any remote instance with incompatible rules, it's even fairly common to silence really large instances like mastodon.social just so they don't drown out the rest of the network if that seems like another issue or fear with Bluesky.

**Comment by devilspeech** [comment]
> Public opt-out bridging has been a part of federation from the beginning a part of the culture... the people complaining about the idea of it are the same kind of people who call the entire ActivityPub Fediverse "Mastodon".

When the initial backlash was happening I admittedly didn't know that it would be necessary for the bridge to essentially follow/copy fediverse accounts en masse for opt out to function, since Bluesky can't see anything on AP otherwise. I'm wondering now if most of the people who were protesting back then were aware of this either, or if there was some miscommunication that happened along the way. Either way the dogpiling was still entirely uncalled for and reflected very poorly on the fediverse as a whole, even if it was only a small number of people doing the worst of it. Fedi culture is not a monolith but the loudest most unruly voices always end up being the ones who leave the biggest impression.

If the bridge from AP -> AT functioned exactly the same way as AT -> AP then I would also want it to be opt out, because then it's ultimately up to the users (or because it's fedi, the instance admins) to decide who gets bridged on an individual basis simply by following each other, or to decide they don't want to be followed and to block each other. That's just a public social platform for you, and also how users connect across different instances on the fediverse by default. I can however understand being hesitant with the way Bridgy has to handle AP -> AT, since even if the users didn't care it would presumably be a tall ask for Bridgy to essentially duplicate the entire fediverse to make opt out work smoothly.

Hopefully I'm understanding all of this correctly anyway, if not feel free to correct me.

**Comment by Tamschi** [comment]
I think that was _part_ of the concern, but the larger issue was in terms of culture and that it would make "all posts" (I think that was a miscommunication) public and searchable. Fortunately the general vibe on Bluesky turned out far kinder than e.g. I expected overall, but that really wasn't that clear at the time.

**Comment by pcottle** [comment]
> The bridge is like any other instance, no data is sent unless users choose to follow each other, and if you don't want to federate you can block the domain at the single click of button. 

I agree with this -- and I'm hoping we can land some kind of instance opt-in (#1305) by the time Threads has more bidirectional Federation support (which is sooner than you might think). I can offer our decision in writing if that helps :P If anything, this would be like "treat us like any other instance, rather than a divergence from how activitypub normally works"

From the Threads side, we are in a federate by default (blocklist mode). The bridge is just any other instance, and I think our users would be super excited to be able to converse with (and see content from) Bluesky.

I actually think opt-in to start was the right choice from @snarfed, especially given the vastly different cultures between the two main platforms at the time and the heated debate that ensued. But now that the dust has settled from the concept of bridges, I think each community can decide what they want.

For Threads, more reach and more cross-instance relationship building is great! I imagine bluesky would feel the same way. We have pretty robust integrity controls so we're not worried as much about a big influx of users (and we are at 275M monthly users anyways).

**Comment by snarfed** [comment]
Thanks for the vote of confidence @pcottle! I'm definitely partial to that perspective on AP federation and bridges too. And we're excited for Threads integration! Following up on instance level opt in for Threads etc on #1305...

**Comment by FurblandChannel** [comment]
I honestly am in support of making it opt-out, I feel like it’s very hard to get people to bridge otherwise. (I know from experience!) Though maybe only create the profiles upon request from a fediverse user, we don’t want random crawls overloading servers.

**Comment by FurblandChannel** [comment]
Actually, you might also want to create profiles upon interaction with a bridged user before then bridging the interaction. Point being, don't just go around making a profile for every account on a server

**Comment by lupomancer** [comment]
Opt-in would be the dream. The bridge is in such a great position now with the ability to set your handle on the BSKY end. 

If the bridge was opt-in, I could wind down my BSKY accounts completely.

I will echo what others said in saying you were dogpiled by a very vocal minority from Fedi. I agree they have good reasons, but I also feel like people who have those worries are also capable of opting out. Anyone else will just benefit from an interconnected network.

Scalability is a huge question though, and I could see that getting out of hand quickly. I think only creating an account on user poll (searching for an account) is a good idea. Mastodon's user base is small, and the users of the bridge are even less. No need to boil the ocean for a small subset of people who will only follow a small subset of BSKY accounts.

As for BSKY's feelings on the matter; I think most users don't even know about the bridge, but would be okay with it if they were just opted in. For most people, any friction at all is too much. Even just following a weird account that they're being asked to follow by a random person.

**Comment by db0** [comment]
I agree with opt-out on bsky. I created a test account for myself and didn't get the DM from mastodon because I didn't realize it blocks PMs by default.

**Comment by Tamschi** [comment]
> Btw, what does one do if the PM they got from the bot is lost because their chat settings hid it? I can't see the chat message in bsky anymore and no new one is being sent.

I'm not sure we can distinguish between users who were sent the DM and those who didn't receive it due to their Chat settings.
That should arrive with #1326, after which you could retry after seeing the error message.

I'm not sure what to do about earlier cases. If we just re-enable it, that would risk sending the request twice to some users.
We should probably expunge seen accounts after a (long) timeout though, at which point requesting would become possible again.

> About opt-out from mastodon users. I know there was a massive blowback the first time, but what if an instance admin could set their whole instance as opt-out through a DNS text record? This way instance admins who are not against it can set it that way.

We haven't decided on a mechanism, but yes that's #1305.

**Comment by therealNAAN** [comment]
> This would likely alleviate a lot of the friction that opt in has introduced. My experience of trying to ask people to opt in on Bluesky has been like pulling teeth, even if everyone has been fine with it once they know it's an option. Most Bluesky users don't have DMs open since that is the default setting (as opposed to fedi where most people don't bother to turn on the setting that blocks DM notifications) so I am essentially forced to have an account on Bluesky to inform people about the bridge. And even then people often don't seem to understand what to do despite my best efforts to explain it in simple terms... Not to mention language barriers, on the off chance that someone does have DMs open the DM that Bridgy sends is also only in English.
> 
> I know it has been discussed ad nauseam, but I hope this could be reevaluated with more perspectives from people who are actually trying to use the bridge now and are running up against these barriers. I have accepted that we just have to live with it on fedi, but it would help us a great deal too if Bluesky didn't have the same artificial restrictions imposed on them.

This, so much! It's very telling that (in my personal experience!) it's been easier to explain to people outside the US about how the bridge works--whereas with Americans, I don't know why, it has been WORSE than pulling teeth! I think I can count on one hand the amount of people in the latter who've understood what I meant. Heck, I've gotten people telling me *I'm* wrong because _Sky Follower Bridge_ already worked for them. And this is after I explicitly say BridgyFed by name! I've accidentally chewed someone's head off after the nth time this has happened. (I apologized later, don't worry, readers at home.)

That said, I would like to mention something that a dear friend of mine said on Bluesky when helping to promote the bridge: anyone on Fedi who is interested in the bridge and has been trying their damnedest to inform Bsky users are the nice people, because anyone who gave everyone else grief over it has already blocked anything BridgyFed on their end. Take that as you will, but I wholeheartedly support making Bluesky opt-out or whatever will make it easier for their accounts to bridge into the Fediverse. Because not going to lie, methinks the users over there are content with never having to do the work we've had to do to keep in touch with our communities and friends who've moved over there...

(Sorry if this ended up being more of a rant than actual helpfulness ^^; My answer is just yes, make it opt-out lol)

**Comment by michael-patchwork** [comment]
> Thanks for the vote of confidence @pcottle! I'm definitely partial to that perspective on AP federation and bridges too. And we're excited for Threads integration! Following up on instance level opt in for Threads etc on #1305...

Great to see this conversation! @snarfed this is the right time to make the bridge opt out on Bsky - yes. In the early days we could go around asking people to bridge, but with so many joining, including many people Fedi users would like to follow, Bsky opt out is the way to go. Also you've now made everything more stable. All subject to getting the governance sorted as you've flagged.

On the Fedi side, instance opt in would be great. So that's a yes to both.

**Comment by FurblandChannel** [comment]
I think we pretty much have a unanimous decision now: bsky bridging should be opt out!

**Comment by pcottle** [comment]
@snarfed thoughts on making a post on bluesky to solicit opinions? Maybe posting from the bridgy fed account + soliciting reposts?

I just want to know what the wider community thinks (even if there is some selection bias)

**Comment by mackuba** [comment]
> This would likely alleviate a lot of the friction that opt in has introduced. My experience of trying to ask people to opt in on Bluesky has been like pulling teeth (…)

Yeah, I think you've more or less described my experience with trying to ask people on Mastodon to enable the bridge too ;)

**Comment by snarfed** [comment]
> @snarfed thoughts on making a post on bluesky to solicit opinions? Maybe posting from the bridgy fed account + soliciting reposts?

Yes! Definitely the right idea. I'll admit that I'm a little gun-shy, but whatever, still worthwhile. Will do!

**Comment by mackuba** [comment]
> When the initial backlash was happening I admittedly didn't know that it would be necessary for the bridge to essentially follow/copy fediverse accounts en masse for opt out to function, since Bluesky can't see anything on AP otherwise.

@snarfed, just to clarify - as I understand, the initial pre-GitHub-thread-drama plan was that no Mastodon account would have a bridged copy created on AT side automatically just when the bridge is launched, but that you'd DM the bot on Bluesky like right now, "hey bot I want to follow snarfed@mastodon.social", and then that account would be bridged at that point for me, without having to first ask the Mastodon user to enable it? Is that right?

**Comment by Tamschi** [comment]
There's one safety aspect I'd like to see with out-out: Allow only bridged followers unless the Bluesky account opts in explicitly.

If we don't do that, then Bluesky users won't have fully native control over their followers, which is a bit of a problem (#1406) since we also can't reliably notify them of off-network followers either.

**Comment by snarfed** [comment]
> @snarfed, just to clarify - as I understand, the initial pre-GitHub-thread-drama plan was that no Mastodon account would have a bridged copy created on AT side automatically just when the bridge is launched, but that you'd DM the bot on Bluesky like right now, "hey bot I want to follow [snarfed@mastodon.social](mailto:snarfed@mastodon.social)", and then that account would be bridged at that point for me, without having to first ask the Mastodon user to enable it? Is that right?

Right!

**Comment by rossabaker** [comment]
Blocking individuals on public feeds may seem futile, easily defeated by alts and logged-out views.  But beyond the minor nuisance to the blocked, it's an explicit withdrawal of consent by the blocker.  Bridging these expressions of consent is just as important as bridging our expressions of creativity.  As a Fedi-to-Bsky user, I would gain a lot from this proposal.  But until we can bridge both types of expression, such as with #1406, I think the opt-in remains the right approach.

**Comment by EmilJacobs** [comment]
Opt-out as the default would make the bridge *so* much more useful. I currently have 500 followers on my bridged account at Bluesky, but only a small fraction, about 25, are visible to me on Mastodon. I give this idea a yay!

**Comment by praichor** [comment]
Unfortunately, opt-in ultimately makes the bridge unusable in practice. As a Mastodon user, I cannot persuade everyone I want to follow on Bluesky to turn on the bridge. In particular, it is impossible to get the attention of celebitry users.

The service is only really useful with opt-out.

And how cool is this bridge!

I also cannot understand the arguments against opt-out: On both sides of the bridge, the basic idea of ​​federation is that the data can be accessed by any other server in the federation, not just current but also future ones. In other words, the data is completely public anyway. You agree to this as soon as you register with a federating social network.

Theoretical question: Would you disapprove, would there a reason against it or would there be anything stopping me from simply hosting a modified opt-out branch of bridgy-fed with a new domain myself? Whoever on Mastodon has something against it can then simply block the domain I use.

(Repost of https://github.com/snarfed/bridgy-fed/issues/1305#issuecomment-2482737419)

**Comment by nileane** [comment]
Yeah, I fully agree that making the bridge opt-out for Bluesky users is the way to go to make it useful to wider range of users on both sides of the bridge.

**Comment by pcottle** [comment]
> the basic idea of ​​federation is that the data can be accessed by any other server in the federation, not just current but also future ones.

I agree that the openness of Federation doesn't match the opt-in behavior of the bridge. I just think there was a huge culture divide between bluesky and mastodon, which is why opt-in was the right choice

> Would you disapprove, would there a reason against it or would there be anything stopping me from simply hosting a modified opt-out branch of bridgy-fed with a new domain myself? 

The confusing thing here is that I think every bluesky user and every activitypub user would be *duplicated* once per bridge, since you'd be on another domain. Not entirely sure of that, but I think having just one bridge is better than having multiple (when they will all be piping the same content)

**Comment by praichor** [comment]
> > the basic idea of ​​federation is that the data can be accessed by any other server in the federation, not just current but also future ones.
> 
> I agree that the openness of Federation doesn't match the opt-in behavior of the bridge. I just think there was a huge culture divide between bluesky and mastodon, which is why opt-in was the right choice
> 

I haven't followed it closely, but wasn't the resistance to opt-out exclusively on the AP side? In other words, no obstacle to Bluesky opt-out?

> > Would you disapprove, would there a reason against it or would there be anything stopping me from simply hosting a modified opt-out branch of bridgy-fed with a new domain myself?
> 
> The confusing thing here is that I think every bluesky user and every activitypub user would be _duplicated_ once per bridge, since you'd be on another domain. Not entirely sure of that, but I think having just one bridge is better than having multiple (when they will all be piping the same content)

Without having thought it through in detail, but wouldn't duplicate bridge accounts only exist if people explicitly used both bridges?

**Comment by snarfed** [comment]
You all are definitely right about multiple bridges causing duplicate accounts. We have this problem now! https://fietkau.software/pinhole , https://rss-parrot.net/ , etc. We've talked about a lot of possible technical approaches in #800, but afaik no other bridge developers have seriously taken me up on it and jumped into design or development there yet.

**Comment by snarfed** [comment]
Also, another angle on the Bluesky opt out question: there's a lot of interest in _instance admins_ being able to turn the bridge on automatically, ie make it opt out, for everyone on their instance: #1305. Flipboard, Threads, and a number of Mastodon instances are interested in doing this.

As discussed here earlier, In many ways, I see the question of whether to make Bluesky opt out similarly. ATProto is a decentralized network, but the Bluesky team owns and administers the biggest instance by far, and they all explicitly want tools like Bridgy Fed to be opt out. I'd of course make it as easy as possible for users to opt out, but for Bluesky users who disagree with that stance by their admins, that might be enough motivation for them to move to a different PDS (ie instance), just like users in the fediverse can take admin approaches into account when choosing their instance.

**Comment by tesaguri** [comment]
While I agree that making bridges from Bluesky opt out, I feel like it needs more precise justification than "making something publicly accessible entails permission for bridge." There are plenty of nuances outside of the dichotomy of "public vs. private", like Mastodon's Quiet Public and `discoverable`, FEP-5feb/FEP-268d search consents and RFC 9309 (Robots Exclusion Protocol), most of which don't have direct counterpart concept on Bluesky and thus make ActivityPub-to-AT-Protocol bridges highly debatable.

So, why are bridges from Bluesky considered okay? Here's how I understand the rationale: In the world of AT Protocol, or the Atmosphere, the users' posts are distributed by servers (Personal Data Servers), typically in response to crawling requests by "relays" for aggregation by arbitrary "AppViews" subscribing to the relays (which include Bluesky). This is in contrast to ActivityPub's architecture where servers have a moderate level of control over how their users' activities are delivered.

In Atmosphere, the common practice is that the posts' schemata ("lexicons") are documenting how they are expected to be aggregated and presented, and the Bluesky lexicon doesn't have much "nuances" like in the Fediverse, with the exception of the logged-in visibility (`!no-unauthenticated` label, resolved in #1098) and reply controls (tracked in #1098, blocked on standardization efforts on the Fediverse's side). So, keeping in mind those limited set of signaling mechanisms, BF is not very different from other AppViews in that it just aggregates and presents the posts according to the schema.

**Comment by michael-patchwork** [comment]
> Also, another angle on the Bluesky opt out question: there's a lot of interest in _instance admins_ being able to turn the bridge on automatically, ie make it opt out, for everyone on their instance: #1305. Flipboard, Threads, and a number of Mastodon instances are interested in doing this.
> 
> 
> 
> As discussed here earlier, in many ways, the question of whether to make Bluesky opt out feels similar. ATProto is a decentralized network, the Bluesky team just owns and administers the biggest instance, and they all explicitly want tools like Bridgy Fed to be opt out. I'd of course make it as easy as possible for users to opt out, but for Bluesky users who disagree with that stance by their admins, that might be enough motivation for them to move to a different PDS (ie instance), just like users in the fediverse can take admin approaches into account when choosing their instance.

Great point!

**Comment by pcottle** [comment]
Yeah great points! I wonder if we can land on these two decisions:

- ActivityPub instances, given the dynamics of the protocol, should have instance-level opt-in controls (#1305). Threads would sign up in a heartbeat, but it would avoid a lot of thrash there.
- ATProto is a decentralized protocol and this bridge similar to any other app running on ATProto. Obviously would be worth giving the team at Bluesky a heads up, but I find Ryan's arguments above solid that they *want* experimentation and developers to build and the ecosystem to be open.

**Comment by praichor** [comment]
> Yeah great points! I wonder if we can land on these two decisions:
> 
>     * ActivityPub instances, given the dynamics of the protocol, should have instance-level opt-in controls ([Instance Opt-in #1305](https://github.com/snarfed/bridgy-fed/issues/1305)). Threads would sign up in a heartbeat, but it would avoid a lot of thrash there.
> 
>     * ATProto is a decentralized protocol and this bridge similar to any other app running on ATProto. Obviously would be worth giving the team at Bluesky a heads up, but I find Ryan's arguments above solid that they _want_ experimentation and developers to build and the ecosystem to be open.

Sounds great:

Even the more cautious Mastodon server admins shouldn't have a problem with this, as opt-in applies to them.

But someone who wants to use one account to operate uniformly on Mastodon, Threads and Bluesky and doesn't have the option to do so today can do so in the future with a Mastodon or Threads account.

(Ignoring the fact that federating in threads is unfortunately opt-in, but that's another topic.)

**Comment by Tamschi** [comment]
> > the Bluesky team just owns and administers the biggest instance, and they all explicitly want tools like Bridgy Fed to be opt out.
> 
> I was not aware of that. So they've explicitly requested this?

It was more of a general statement regarding compatible services/bridges I believe (though within a conversation about Bridgy Fed), but I don't have it on hand and could misremember.

In any case, I think the main blocker for Bluesky-side opt-out right now isn't whether this would be acceptable socially (I suspect so) or map well at a protocol level (interaction controls wouldn't map well yet, though Bluesky does filter out almost all effects of disallowed interactions. Bridgy Fed could also choose to specifically not bridge posts that have interaction controls on them.), but that Bridgy Fed as a project isn't necessarily ready to make these decisions: [Possible futures for Bridgy Fed](https://snarfed.org/2024-11-01_53932)

**Comment by snarfed** [comment]
@Tamschi is right on all counts! And sorry for the confusion, I think I added the _now_ label to this issue just to remind me to make the public post to request feedback that @pcottle mentioned in https://github.com/snarfed/bridgy-fed/issues/1471#issuecomment-2478771409, not to mean that I'm actively planning for or working on making Bluesky opt out yet. That's still blocked on real organization structure, maybe funding, etc.

**Comment by snarfed** [comment]
@priachor, re:

> Theoretical question: Would you disapprove, would there a reason against it or would there be anything stopping me from simply hosting a modified opt-out branch of bridgy-fed with a new domain myself? Whoever on Mastodon has something against it can then simply block the domain I use.

Hah, yes! The agony and ecstasy of open source. It's public domain licensed, so there's definitely nothing stopping you. I have no illusion that I can control or (necessarily) influence what other people do with my open source code.

I'd love to see other bridges! If we started seeing a lot more, I think we'd need to seriously tackle #800, which is possible.

Otherwise, Bridgy Fed itself isn't designed to be easily self-hosted, it's somewhat tied to Google Cloud, but it's very doable to run another instance of it there. It's also not designed to be easily "white labeled," yet, so if anyone does run a new instance, I'd strongly encourage them to rename it and otherwise update the front page and docs to make it clear that it's a different service, run by someone else.

Regardless, happy to help if you want to try!

**Comment by Fauli1221** [comment]
@snarfed why was this tagged with the now tag by the way?
are you actually planning or rolling it out soon?

**Comment by snarfed** [comment]
@Fauli1221 https://github.com/snarfed/bridgy-fed/issues/1471#issuecomment-2484097245

**Comment by leo60228** [comment]
This is more or less how I expected and hoped Bridgy Fed to work when I heard that someone was working on an atproto/ActivityPub bridge over a year ago, for what it's worth.

Two other factors that make this especially useful:

* Many of the Bluesky accounts I'd like to follow are large accounts that would be difficult to get in touch with and request to enable the bridge.
* I've noticed that people who don't themselves have the bridge enabled interact with my bridged posts. I have to manually watch for this, and might miss replies to old posts.
  * This is #1200, and even if the opt-out change is made it would admittedly be ideal to still have *some* solution for delivering notifications about interactions from opted-out accounts.

**Comment by EdlReg** [comment]
I am a journalist, and from my perspective I think anyone interested in promoting the Fediverse should strive for making Bridgy opt-out on Bluesky – and well, other serious networks as well. 
Many large news organizations will now be establishing a presence on Bluesky. But that is only step one, and if they notice a large audience in the Fediverse as well (through a bridge), they might – maybe in a year or so – start thinking of creating their own instances. So: for Mastodon and other large actors in the Fediverse this should be strategically important. 
I do understand that you need resources and a larger group working with this, @snarfed. Would it be possible to integrate Bridgy more directly with Mastodon, and get money from grants etc? Just thinking aloud.

**Comment by ronilaukkarinen** [comment]
I vote for opt-out. It's near impossible to get users to use the Bridgy Fed now that it is opt-in. Many want to, but are not tech-savvy. I agree that BrdigyFed is like any other instance you can block or follow and you should have a way to disable your posts to be seen by other users.

I've been harassed by vocal users about adding their Mastodon account to my personal private list or trying to find their otherwise completely public post, or implemnting a public text search opt-out by VyrCossont on my server (which later was implemented to Mastodon core as opt-out). This is the level we are on [here](https://github.com/snarfed/bridgy-fed/issues/835), that it would be somehow illegal to see one's public post elsewhere on a federated network.

Public posts are public.

This is my take as an instance admin. I vote yes for making BridgyFed opt-out instead of opt-in.

**Comment by EmilJacobs** [comment]
Any updates on this? Since the bridge is up and running again, it would be great if we moved on this. Maybe as a question: is there anything holding this up?

**Comment by snarfed** [comment]
From the issue description:

> I wouldn't really consider making this change until if/when Bridgy Fed has some kind of real organizational structure, governance, and sustainability, as opposed to now, where it's just one person's side project.

Not the only blocker here, but probably the biggest. Fortunately we're working on it!

**Comment by praichor** [comment]
> Fortunately we're working on it!

Sounds very exciting! Is there more information you can share?

**Comment by snarfed** [comment]
Not yet. Stay tuned!

**Comment by praichor** [comment]
BTW, thank you for your work! Sometimes it feels like the free internet is based on a few key hobby projects by idealists.

**Comment by jonpincus** [comment]
Agreed that norms are different on Bluesky, but even though it's an all-public environment, people there reacted very strongly to the Hugging Face uploads that Jay had encouraged -- much more than I would have expected, tbh, and enough so that the Hugging Face researcher took down the first data set.  So it's probably worth seeking other input; I'm not sure how representative the feedback here is.

@pcottle when you talk about Threads wanting an opt-out model for Bridgy Fed, is that in addition to or instead of of Threads' existing opt-in to federation?

(And I have to say it's really disappointing to see people here still arguing that the default in general for fedi in general should be opt-out.)

**Comment by praichor** [comment]
I think it is important to distinguish between:

1. The forwarding of posts within a Fediverse in the broadest sense. 

   As soon as I publish something in a Fediverse, I implicitly give permission for it to be distributed without restriction in the Fediverse. Because that is the core purpose of the Fediverse.

   In other words: opt-out, of course. Anything else makes no sense.

   Analogy: If I publish a website, then I implicitly give permission for visitors' browsers to load the content into the computer's memory and display it.

   Threads is an exception and uses opt-in for the sole reason that most Threads users do not see themselves as part of a Fediverse, but within a closed garden. Bluesky/atproto and Mastodon/activitypub, on the other hand, are heavily promoting the idea of ​​establishing a Fediverse in which other applications should also participate, and therefore opt-out is of course the only right thing to do.

4. The use of the data for other purposes.

   Just because you publish something doesn't mean you have given permission for all types of use.

   For example, for AI training. Example: Hugging Face.

   Analogy: Even if I publish something on a website, it may still be protected by copyright, and of course I don't give permission for unrestricted use.

**Comment by stinerman** [comment]
>     2. The use of the data for other purposes.
>        Just because you publish something doesn't mean you have given permission for all types of use.
>        For example, for AI training. Example: Hugging Face.
>        Analogy: Even if I publish something on a website, it may still be protected by copyright, and of course I don't give permission for unrestricted use.

In practice the only way to stop someone from AI training on your posts or infringing your copyright is to get lawyers involved, and ultimately, bring a suit.

If someone is concerned about this, their best bet is to turn on AUTHORIZED_FETCH in Mastodon. I am not certain if other Fediverse applications have a similar setting. Failing that, the account needs to post with visibility only for followers or some other restriction.

Saying "I want my posts accessible by anyone with an internet connection, but to only be used in some ways" is effectively impossible.

**Comment by lucajet** [comment]
exactly, in practice is 
"either you post just for followers" or, "it's public and the world is full of bad guys"

Post accordingly

**Comment by shiribailem** [comment]
Y'all, lay off please. This is a dead horse of an argument and all of these points have been made.

Ryan has chosen to stick with opt-in on the AP side, and opt-out on the Bluesky side is pending bigger organization around the bridge. Continuing to hammer on it is going to do nothing but wear out Ryan's patience for the whole project.

It's settled to the point that it's no longer a logic issue, it's people dogpiled Ryan and they opted to concede to their demands, not because the demands were good but because it was not a fight Ryan wanted to pursue.

If you want opt-out on the AP side, at this point your option is to fork the bridge, modify it to change it to opt-out and then run your own bridge.

**Comment by FurblandChannel** [comment]
> Y'all, lay off please. This is a dead horse of an argument and all of these points have been made.
> 
> Ryan has chosen to stick with opt-in on the AP side, and opt-out on the Bluesky side is pending bigger organization around the bridge. Continuing to hammer on it is going to do nothing but wear out Ryan's patience for the whole project.
> 
> It's settled to the point that it's no longer a logic issue, it's people dogpiled Ryan and they opted to concede to their demands, not because the demands were good but because it was not a fight Ryan wanted to pursue.
> 
> If you want opt-out on the AP side, at this point your option is to fork the bridge, modify it to change it to opt-out and then run your own bridge.

Agreed that the bridge should remain as it is for now. We don't want a repeat of [this](https://fba.ryona.agency/?domain=fed.brid.gy).

**Comment by praichor** [comment]
> Agreed that the bridge should remain as it is for now. We don't want a repeat of [this](https://fba.ryona.agency/?domain=fed.brid.gy).

Why not?

It is the right of every user to block a domain. And it is the right of every server admin to block a domain. But why should that make fed.brid.gy unusable for everyone else?

**Comment by lucajet** [comment]
I thought we were eventually asking for the bridge to be opt-out from the BS side, not the other way around.

Anyway, 
i don't want to be rude but, with all due respect, those listed instances, with those names, with those well thought technical reasons

it's not exactly like we are speaking about having the adult admins of the top 10 instances against

**Comment by pcottle** [comment]
> @pcottle when you talk about Threads wanting an opt-out model for Bridgy Fed, is that in addition to or instead of of Threads' existing opt-in to federation?

It'd be opt-in for federation, but once you're federated, you're on the open social web and that includes bridges (imo). But I understand if we want to settle on BS being opt-out and ActivityPub being opt-in for now

@praichor I agree with your distinction between the two as well. Bluesky is a lot more permanent and distributed than most people realize (including public blocks!) and I'd rather set user expectations appropriately vs have these surprises later

**Comment by lucajet** [comment]
@pcottle any news about activitypub in threads for EU users ?
I'd need a bridge between mastodon and threads to follow people there !

**Comment by pcottle** [comment]
@lucajet commented about this a while back here:
https://www.threads.net/@pcottle/post/DCMnkLqvdCt
<img width="631" alt="Screenshot 2024-12-04 at 9 44 05 AM" src="https://github.com/user-attachments/assets/acb7b85e-6f84-406f-b30b-777dbaa220c4">

It's not only our decision but also something we work through with EU regulators, who are very busy 😇

**Comment by Hedwig7s** [comment]
Alright so after going over this a bit I'm a little bit confused
From what I've gathered bluesky will be opt-out in the future but AP will stay opt-in. What is the reason for AP staying opt-in exactly? Is it a technical issue? A culture issue?

**Comment by Jay2645** [comment]
> Alright so after going over this a bit I'm a little bit confused From what I've gathered bluesky will be opt-out in the future but AP will stay opt-in. What is the reason for AP staying opt-in exactly? Is it a technical issue? A culture issue?

It's a cultural issue; Mastodon folks are very privacy-conscious and make a BIG deal about stuff mirroring Mastodon in any form. They threw a fit about Mastodon implementing a basic post-based search, for example.

See also: their reaction to Threads being based on the fediverse.

**Comment by osmaa** [comment]
Chiming in to voice my support for opt-out bsky->ap, but the are more nuanced in-between options. The request-to-follow model exists in AP (it does not exist on bsky) meaning someone who wants to control who follows them can have explicit control. The @bsky.brid.gy bot could base its consent on a follow request, not bridging posts before the fedi user approves and follows back. That's still an explicit opt-in, but one with a better notification model. Don't like to be bridged? Don't approve the follow request, or block the bridge.

On bsky side the request-to-follow is missing. However any interaction with a bridged account could be argued to effect opt-in -- it's after all an explicit action by the bsky user which they expect to notify the account on the other side (whether of just a like, or specifically in case it's a reply). Non-delivery of those messages due to a lack of opt-in by the bsky side user is effectively breaking two contracts: 
1. the bridged AP user isn't being notified of engagement with their content, and 
2. the bsky user isn't being told that their interaction doesn't reach the account they intended it for. 

It's as if an unrelated third party (the bridge) implements an unconsented mute/block in between.

Just like today, this minor policy modification would mean that a bsky user who has never interacted with a bridged AP account would be invisible to AP, but the first interaction would implement the bridge. It would simply widen the "first interaction" from you-need-to-follow-@ap.brid.gy to any interaction with the account you wanted to interact with, ie what probably is considered the natural action anyway.

**Comment by lucajet** [comment]
very good points from @osmaa 

theoretically someone in bsky can think i'm rude because i never answered to some interactions i didn't see.

Also, not having a real account there i have a question:
Is there the possibility someone tryto send me a DM or do we correctly appear like people not accepting dm ?

**Comment by Tamschi** [comment]
> Chiming in to voice my support for opt-out bsky->ap, but also that there are more nuanced in-between options. The request-to-follow model exists in AP (it does not exist on bsky) meaning someone who wants to control who follows them can have explicit control. The @bsky.brid.gy bot could base its consent on a follow request, not bridging posts before the fedi user approves and follows back. That's still an explicit opt-in, but one with a better notification model. Don't like to be bridged? Don't approve the follow request, or block the bridge.

That doesn't really work, unfortunately, for a range of technical reasons.
We can't send a "follow request" for example, just a `Follow` (which is auto-approved in most cases).

(The follow-request from a service account would also be counterproductive product design, in my opinion.
It would most likely have extremely low uptake and is likely to make people explicitly opt out of bridging altogether, and less likely to reconsider when asked about it through non-automated channels.)

There's also a **very** large technical discrepancy in privacy controls between Bluesky and ActivityPub, to the point where only the absolute highest multi-opt-in publicity level in Mastodon is comparable to the _lowest_ publicity on Bluesky. We definitely can't consider any accounts who don't have that for auto-bridging, which would limit any opt-out approach to a small fraction of the user base on specific software that supports these opt-ins.

**Comment by Tamschi** [comment]
> Also, not having a real account there i have a question: Is there the possibility someone tryto send me a DM or do we correctly appear like people not accepting dm ?

Bridge accounts on Bluesky have their Chat fully closed, including to Chat requests by their follows.

We can't do the same on the ActivityPub side, unfortunately. I hope someone comes up with a scheme similar to [GoToSocial's Interaction Policies](https://docs.gotosocial.org/en/latest/federation/posts/#interaction-policy) there, since those are only applicable to posts. (It's definitely not trivial since addressing in ActivityPub has practically unlimited granularity.)

**Comment by Hedwig7s** [comment]
> > Alright so after going over this a bit I'm a little bit confused From what I've gathered bluesky will be opt-out in the future but AP will stay opt-in. What is the reason for AP staying opt-in exactly? Is it a technical issue? A culture issue?
> 
> It's a cultural issue; Mastodon folks are very privacy-conscious and make a BIG deal about stuff mirroring Mastodon in any form. They threw a fit about Mastodon implementing a basic post-based search, for example.
> 
> See also: their reaction to Threads being based on the fediverse.

If I'm being honest not this project's problem in the slightest
Don't like it? Opt out or find an instance that blocks it your posts are public deal with it

**Comment by Tamschi** [comment]
> > > Alright so after going over this a bit I'm a little bit confused From what I've gathered bluesky will be opt-out in the future but AP will stay opt-in. What is the reason for AP staying opt-in exactly? Is it a technical issue? A culture issue?
> > 
> > 
> > It's a cultural issue; Mastodon folks are very privacy-conscious and make a BIG deal about stuff mirroring Mastodon in any form. They threw a fit about Mastodon implementing a basic post-based search, for example.
> > See also: their reaction to Threads being based on the fediverse.
> 
> If I'm being honest not this project's problem in the slightest Don't like it? Opt out or find an instance that blocks it your posts are public deal with it

I'm sorry, but I think I need to shut down this angle a bit in general (regarding the primary instance of Bridgy Fed, though).

Communication services are only useful if they actually are used by people. A surefire way to _not_ achieve that is to ignore public image of the service and preferences of prospective users. So it very much _is_ our problem (or rather: something we _want_ to take into account, in part in order to make Bridgy Fed as broadly useful as feasible and, at least in _my_ opinion, because it's the right thing to do in situations like this).

You may have a different preferences individually and that is fine, but it doesn't override others' preferences and expectations.
Technology in general doesn't exist in a vacuum and most practical software design problems are rooted in human interaction design and impact analyses, not technical aspects or technological feasibility. It's not the protocols that allow us to have this service here, it's users' and admins' acceptance of them and what they consider it means to be a beneficial participant in each network.

**Comment by lucajet** [comment]
> > Also, not having a real account there i have a question: Is there the possibility someone tryto send me a DM or do we correctly appear like people not accepting dm ?
> 
> Bridge accounts on Bluesky have their Chat fully closed, including to Chat requests by their follows.

if they can't even try to, is totally fine by me, they won't be tricked into thinking they can

**Comment by Tamschi** [comment]
> > > Also, not having a real account there i have a question: Is there the possibility someone tryto send me a DM or do we correctly appear like people not accepting dm ?
> > 
> > 
> > Bridge accounts on Bluesky have their Chat fully closed, including to Chat requests by their follows.
> 
> if they can't even try to, is totally fine by me, they won't be tricked into thinking they can

Bluesky shows a Chat button on profiles of users that you can message: 
![speech bubble button with three dots](https://github.com/user-attachments/assets/0eba558a-aab3-4538-b46e-aec3e402fa00)

This button is always missing for bridge accounts and the "Start a new chat" dialogue in the Chat UI also shows these users greyed-out and unselectable.

**Comment by Hedwig7s** [comment]
> > > > Alright so after going over this a bit I'm a little bit confused From what I've gathered bluesky will be opt-out in the future but AP will stay opt-in. What is the reason for AP staying opt-in exactly? Is it a technical issue? A culture issue?
> > > 
> > > 
> > > It's a cultural issue; Mastodon folks are very privacy-conscious and make a BIG deal about stuff mirroring Mastodon in any form. They threw a fit about Mastodon implementing a basic post-based search, for example.
> > > See also: their reaction to Threads being based on the fediverse.
> > 
> > 
> > If I'm being honest not this project's problem in the slightest Don't like it? Opt out or find an instance that blocks it your posts are public deal with it
> 
> I'm sorry, but I think I need to shut down this angle a bit in general (regarding the primary instance of Bridgy Fed, that is).
> 
> Communication services are only useful if they actually are used by people. A surefire way to _not_ achieve that is to ignore public image of the service and preferences of prospective users. So it very much _is_ our problem (or rather: something we _want_ to take into account, in part in order to make Bridgy Fed as broadly useful as feasible and, at least in _my_ opinion, because it's the right thing to do in situations like this).
> 
> You may have different preferences individually and that is fine, but it doesn't override others' preferences and expectations. Technology in general doesn't exist in a vacuum and most practical software design problems are rooted in human interaction design and impact analyses, not technical aspects or technological feasibility. It's not (just) the protocols that allow us to have this service here, it's users' and admins' acceptance of them and what they consider it means to be a beneficial participant in each network.

If I'm being honest it is almost definitely a loud minority complaining about their "privacy" (on a public platform) because people who aren't opposed won't be saying anything
I feel most mastodon users probably do not care (it's not much different compared to any other instance)

**Comment by lucajet** [comment]
I heard of a forthcoming bridge between nostr and bluesky by Alex Gleason.
It will be interesting to see if they foresee an opt-in mechanism BS side, and if not, if will be blocked and how quickly.

**Comment by Tamschi** [comment]
For what it's worth, we have #1305 to make specific instances opt-out. That's something the admins there should decide though.

---

> I heard of a forthcoming bridge between nostr and bluesky by Alex Gleason.
> It will be interesting to see if they foresee an opt-in mechanism BS side, and if not, if will be blocked and how quickly.

Bridging content towards Nostr is probably not much of an issue. Content easily appearing elsewhere is something that the Bluesky team wants, I think. I'm curious to see how Bluesky handles incoming Nostr content though, if the bridge itself isn't moderated.

**Comment by osmaa** [comment]
> Most users are lurkers who follow large accounts, for example https://universeodon.com/@georgetakei.
> 
> A key use case is therefore that you can follow large, well-known users from the other network. However, such accounts do not respond to individual requests and there will be no "first interaction".

George Takei is probably a bad example, because he is separately very active at https://bsky.app/profile/georgetakei.bsky.social. He may intentionally choose to keep those separate for reasons that are not for me to speculate. That said, someone seeking a large audience is hard to real in-band, but would hopefully get the message about opt-in out-of-band somehow - trusted contacts and so forth.

Reach is also about reaching the small accounts and individuals, and convenience and in-band message is more relevant there. That's where keying into the first interaction should help.

> That doesn't really work, unfortunately, for a range of technical reasons.
> We can't send a "follow request" for example, just a Follow (which is auto-approved in most cases).

Right, sorry, I should have checked what the protocol actually does. AP follow model is anyway two-phase (before the follow approval, auto or manual, no messages are sent), where bsky isn't. 

> It would most likely have extremely low conversion rates, in marketing terms, and is likely to make people explicitly opt out of bridging altogether instead

The current model is extremely low conversion rate. I think that's what this whole opt-in vs opt-out discussion is out to solve, is it not?

> Communication services are only useful if they actually are used by people. A surefire way to not achieve that is to ignore public image of the service and preferences of prospective users. 

This I couldn't agree more on! A solution which doesn't meet public expectation will fail. That said, no solution will make everyone happy - there's always a minority who is unhappy with any choice.

Anyway, thanks for all the feedback! I learned something new today reading this through.

**Comment by shiribailem** [comment]
> This I couldn't agree more on! A solution which doesn't meet public expectation will fail. That said, no solution will make everyone happy - there's always a minority who is unhappy with any choice.

That's honestly been my entire point about opt-in vs opt-out on the AP side, the people who were screaming at you are 99% the people who are unhappy regardless. Practically none of them use the bridge in the first place, the instance admins who view things that way all block the bridge entirely.

My whole complaint has been that the opt-in mechanics exist to appease the people who will never use it to begin with, people who will never be happy to begin with. These are largely the same people who became angry and started the campaign to try and block Threads everywhere because in their mind it's some sort of corporate invasion for them to... support a protocol.

**Comment by TomCasavant** [comment]
> You may have different preferences individually and that is fine, but it doesn't override others' preferences and expectations.

It's probably been discussed somewhere in one of the *many* different threads for this discussion haha, but my problem with the opt-in approach is it _does_ seem to break expectations for both sides of the bridge because you eliminate the ability to block/mute accounts that haven't opted in (though that may be fixed with whatever solution is decided on for #1406). If you make it opt out, then anyone who intentionally blocks the bridge wouldn't be able to interact with bridged users (and so you wouldn't need to block unbridged users). 

And similarly, though less of a risk, people expect that if they reply or like a post that the poster gets notified of that activity- and anyone unaware of the bridge will be unaware that their message was not delivered. Notifying users that they replied to a bridged post helps in cases where users have their DMs turned on/public, but outside of those situations they would never know.

But, I think for the most part it's already been decided that it'll be opt-out for the non-Threads ActivityPub side, which is fine- just putting my thoughts out there.

**Comment by snarfed** [comment]
> There's also a **very** large technical discrepancy in privacy controls between Bluesky and ActivityPub, to the point where only the absolute highest multi-opt-in publicity level in Mastodon is comparable to the _lowest_ publicity on Bluesky. We definitely can't consider any accounts who don't have that for auto-bridging, which would limit any opt-out approach to a small fraction of the user base on specific software that supports these opt-ins.

@Tamschi are you mainly thinking of audience targeting for posts and other AP objects here? ie public, unlisted, followers only, etc?

The only account-level controls I'm aware of are `discoverable` and `indexable`, which BF currently sees and understands, but lets fediverse users override if they choose to opt into the bridge. Are there others?

**Comment by Tamschi** [comment]
> > There's also a **very** large technical discrepancy in privacy controls between Bluesky and ActivityPub, to the point where only the absolute highest multi-opt-in publicity level in Mastodon is comparable to the _lowest_ publicity on Bluesky. We definitely can't consider any accounts who don't have that for auto-bridging, which would limit any opt-out approach to a small fraction of the user base on specific software that supports these opt-ins.
> 
> @Tamschi are you mainly thinking of audience targeting for posts and other AP objects here? ie public, unlisted, followers only, etc?
> 
> The only account-level controls I'm aware of are `discoverable` and `indexable`, which BF currently sees and understands, but lets fediverse users override if they choose to opt into the bridge. Are there others?

It's only those, yes. (Discovery and indexing are the two major differences where someone can't opt out on Bluesky.)

**Comment by jonpincus** [comment]
> Opt-in of a bridge forces every user to use a methodology that makes the service virtually unusable for all. Opt-out gives everyone the freedom to decide for themselves.

Opt-in also gives everyone the freedom to decide for themselves. For consent to be meaningful, it needs to be a freely given, informed, affirmative act. Opt-out assumes consent without the person making an affirmative act or necessarily being informed.  As a result, opt-out doesn't protect people who are potentially at risk.

It's a fair point that Bluesky's current model is non-consensual data sharing, and leaving it up to people who are potentially at risk to protect themselves, so I can certainly see the argument that Bridgy Fed should have an opt-out model for Bluesky users.  Still, the reaction to the Hugging Face incident implies that many users don't like Bluesky's non-consensual approach, enough that Bluesky's at least saying that they're considering implementing an (advisory) consent mechanism.  So I can also see the argument for keeping it opt-in on the Bluesky side.

**Comment by leo60228** [comment]
I don't think you can reasonably argue that there isn't a fundamental difference between mass scraping and protocol bridging. Using Bluesky fundamentally requires granting consent for your data to travel between servers.

**Comment by AtiusAmy** [comment]
Is there any update on this after the organization news?

**Comment by rwmpelstilzchen** [comment]
Organization news?
@AtiusAmy

**Comment by AtiusAmy** [comment]
> Organization news? [@AtiusAmy](https://github.com/AtiusAmy)

Referring to this: https://snarfed.org/2024-11-01_53932

and afterwards we got https://www.anew.social/hello-social-web/

So I was just wondering

**Comment by snarfed** [comment]
@AtiusAmy sorry, no news yet. We'd love to work more on this eventually!

**Comment by Henry-Hiles** [comment]
This would be really nice in order to onboard new people to fedi, so they know that they can still talk to their friends on bsky without explaining to them to follow a bot

**Comment by laurence-b** [comment]
Thanks for linking all related issues, it helps seeing how much of a recurrent theme it is! (My atypical Github user/ less tech-savvy view #1790).

I quote from elsewhere what you said @snarfed "If a platform or network like the fediverse is opt in, and someone has explicitly chosen not to opt in, they'd generally expect that bridge not to deliver their content or profile or interactions anywhere."

I'd just like to add the nuance of 'explicitly chosen not to opt in'. For a tech savvy person, 100%, they understand optin/optout. 

For the average user however, it is not as clear. I liked the suggestion @shiribailem made about echo accounts, or @ch0ccyra1n which both point to the fact that people don't know where accounts live and won't know an account is bridged, or for that matter, not *care* where this account lives. They show however through their behaviour interacting with it, that they expect to be able to do so and not have to worry about the name of the account or something as techie as 'what server does it live on'. 

They are further encouraged to think the bridge is nothing, as they are able to follow the account without any sort of warning when they do so. From a newbie's perspective, it makes zero sense to differentiate these actions, which basically all amount to 'interacting with an account'. 

So for comments, maybe a warning message before something gets posted, along the lines of:
_- you are about to comment on an account that isn't on Bluesky: are you happy for your comment to be visible in the [[Fediverse](https://www.fediverse.to/)] social network?_ YES POST NOW /NO I WON'T COMMENT
_-  (additional tickbox): tick here if you would like all your future comments on [[Fediverse](https://www.fediverse.to/)] posts to get automatically published without needing to re-authorise_

If the latter, they should have a confirmation DM with something like 'Congrats! You are now bridged to the Fediverse! This means that... If you change your mind..."

I am assuming here that consent:
- can be obtained as /before someone post a comment on a bridged account
-  is for the Fediverse, rather than each instance, which would be a terrible pain.

**Comment by lucajet** [comment]
yeah, not only for comments, for following also.
Plenty of followers who, i'm sure don't want to read my stuff without interacting, i'm no news agency

**Comment by Henry-Hiles** [comment]
> Thanks for linking all related issues, it helps seeing how much of a recurrent theme it is! (My atypical Github user/ less tech-savvy view [#1790](https://github.com/snarfed/bridgy-fed/issues/1790)).
> 
> I quote from elsewhere what you said [@snarfed](https://github.com/snarfed) "If a platform or network like the fediverse is opt in, and someone has explicitly chosen not to opt in, they'd generally expect that bridge not to deliver their content or profile or interactions anywhere."
> 
> I'd just like to add the nuance of 'explicitly chosen not to opt in'. For a tech savvy person, 100%, they understand optin/optout.
> 
> For the average user however, it is not as clear. I liked the suggestion [@shiribailem](https://github.com/shiribailem) made about echo accounts, or [@ch0ccyra1n](https://github.com/ch0ccyra1n) which both point to the fact that people don't know where accounts live and won't know an account is bridged, or for that matter, not _care_ where this account lives. They show however through their behaviour interacting with it, that they expect to be able to do so and not have to worry about the name of the account or something as techie as 'what server does it live on'.
> 
> They are further encouraged to think the bridge is nothing, as they are able to follow the account without any sort of warning when they do so. From a newbie's perspective, it makes zero sense to differentiate these actions, which basically all amount to 'interacting with an account'.
> 
> So for comments, maybe a warning message before something gets posted, along the lines of: _- you are about to comment on an account that isn't on Bluesky: are you happy for your comment to be visible in the [[Fediverse](https://www.fediverse.to/)] social network?_ YES POST NOW /NO I WON'T COMMENT _- (additional tickbox): tick here if you would like all your future comments on [[Fediverse](https://www.fediverse.to/)] posts to get automatically published without needing to re-authorise_
> 
> If the latter, they should have a confirmation DM with something like _"Congrats! You are now bridged to the Fediverse! This means that... If you change your mind..."_
> 
> I am assuming here that consent:
> 
> * can be obtained as /before someone post a comment on a bridged account
> * is for the Fediverse, rather than each instance, which would be a terrible pain.

Hi, I don't know much about bsky, but I don't see consent could be asked before the comment is posted. It's not like Bridgy Fed controls the bsky instance.

**Comment by praichor** [comment]
The vast majority of people don't care where their data is stored. They just want a well-functioning app and want to see the messages of the people they are interested in, and some of them want to interact with them. For all these people, anything other than opt-out is simply a bad user experience because they are being restricted for no understandable reason.

And of the minority who are concerned about which company's server the data is stored on, most of them didn't choose Bluesky or Mastodon for nothing, who both advertise that you not locked in with this company. For these people, anything other than an out-out contradicts the philosophy of these platforms.

There is a minority of this minority who want their posts to be forwarded to some other servers, but not to other other servers. These people are vocal advocates of opt-in.

**Comment by jonpincus** [comment]
> There is a minority of this minority who want their posts to be forwarded to some other servers, but not to other other servers. These people are vocal advocates of opt-in.

To the contrary: in my poll last year, 71% preferred opt-in for a bridge to Bluesky.  (And that was before the publicity about Bluesky datasets being posted on Hugging Face, or the ramp-up of anti-trans abuse on Bluesky.)

And this is consistent with attitudes in generally.  In 2023, polls in Washington state showed over 70% support for opt-in data protection.

Consent is popular!  It's extremely popular with women and trans and non-binary people, but a lot of cis guys think consent is a good thing too!

**Comment by praichor** [comment]
> To the contrary: in my poll last year, [71% preferred opt-in for a bridge to Bluesky](https://infosec.exchange/@thenexusofprivacy/112157733279294857).

Well, not exactly a scientific survey with a statistically random selection of all users. That's what I mean by "vocal advocates". I would be cautious about allowing far-reaching decisions of great consequence to be determined by a very small but vocal minority.

> And this is consistent with attitudes in generally. In 2023, polls in Washington state showed over 70% support for opt-in data protection.
> Consent is popular! It's extremely popular with women and trans and non-binary people, but a lot of cis guys think consent is a good thing too!

This is about something completely different. It's about personal and private data. Not data that you specifically make available to the public for distribution to other servers.

You join a Mastodon server precisely for the purpose of having it forward your posts to other servers (or not), according to your server's policy. And even if you don't like the server's policy, but for whatever reason you still want to stay on the server, you can block other servers as much as you want. At least in case the server blocks too little for your liking. If it blocks too much, you have a problem.

**Comment by stacyharper** [comment]
I'm sorry to interrupt, but isn't this about Bluesky opt out? First, and if I'm right, Bluesky posts are public anyway. So basically, the Fediverse just need a translation bridge to access them. In the same time, a Fediverse public post should also be readable through a similar bridge to the Bluesky network. Should we skip the restricted posts from the conversation for now? bridge-fed does not handle them anyway right?

That being said, and in my opinion, Bluesky users never had the choice to restrict their posts visibility. They just posts things that anyone can read. In the contrary, Mastodon/Pleroma/etc users can restrict their posts visibility, and should not be surprised that their public posts can be read from another network.

**Comment by mackuba** [comment]
> But it's not just a matter of people following them from the other instance. Even if Ive chosen the option to approve followers and don't accept anybody from Bluesky, if it was opt-out my posts could potentiall be shared to Bluesky without my knowledge or consent if somebody who _does_ have followers on Bluesky boosts them.

Hm, no, I don't think so? In order for a boost to be bridged (in the form of someone's Bluesky-mirror doing a Bluesky-repost), that post must already have its Bluesky representation, boosting can't create it on demand. And the post will only have a Bluesky copy if at least one person is following that Mastodon account from Bluesky - because Bridgy doesn't "fetch" posts through some fetch API, it just uses ActivityPub following, which works in such way that you only start being sent posts from the author when you follow them.

So the whole thing was just about whether:

- you can be followed by anyone from Bluesky at any time (which you can block/refuse if you notice it or you have follow confirmation enabled), and only *then* your posts start appearing on Bluesky (original idea)
- or, if you need to first opt in in order for people to be able to follow you and make your posts flow to Bluesky (current version)

**Comment by jonpincus** [comment]
Hmm, you're right that I was making an assumpoint -- that's how boosting works in other situations.  

With the current version, your posts flow to Bluesky if you've opted in (whether or not you have any followers). If it had gone the opt-out route, posts would presumably still flow if you've got at least one follower there, but there would have been an implementation choice about whether to bridge a boost (by an account that had at least one Blueksy follower) of a post by an account that hadn't opted out of bridging.

**Comment by praichor** [comment]
> People joining up on Bluesky know that (or should know that); that's the choice they're making. People signing up on Mastodon _aren't_ choosing that.

This is a widely spread misunderstanding of how the Fediverse works. In the Fediverse there is no "joining up on Bluesky" versus "signing up on Mastodon". In the Fediverse (regardless of the technical details of ActivityPub versus ATProto most people don't care about) there are countless different servers and operators with their own, sometimes very different moderation rules, very different cultures. You are signing up to one of very many servers and choose its specific moderation rules and culture.

A lot of education is still needed here. However, a lack of knowledge should not be a reason to unnecessarily restrict the freedom of social networks technically. Instead, every user should have the freedom to decide what should be forwarded, how and where.

Someone who does not want their posts to be forwarded to Bluesky can simply choose a server that they consider secure and which does not do this. It's simple for admins to configure their server:

https://fedi.tips/how-to-defederate-fediblock-a-server-on-mastodon/

You can also use a whitelist approach:

https://docs.joinmastodon.org/admin/config/#limited_federation_mode

Or a user can simply block the server directly. Mastadon makes this easily possible:

https://docs.joinmastodon.org/user/moderating/#block-domain

There is no need to patronize the vast majority of users with unnecessary technical restrictions.

**Comment by Jay2645** [comment]
Rather than having a limit on server size (which I agree makes sense), it may be better to make a toggle for entire instances to be opt-in, which could be done by Mastodon server admins. I thought there was a discussion about that elsewhere already, but I can't find it now.

Regardless - I do think the Mastodon side of things is/should be a completely separate issue. This is specifically making Bluesky itself opt-out instead of opt-in, which IMO would be healthy for the fediverse writ large (as it would make a migration to Mastodon easier when the majority of Bluesky accounts already are bridged).

IMO, all Bluesky accounts should be auto-bridged unless they have the toggle to only allow showing their profile to logged-in users. (I thought we already arrived at this conclusion at some point?)

**Comment by snarfed** [comment]
> Rather than having a limit on server size (which I agree makes sense), it may be better to make a toggle for entire instances to be opt-in, which could be done by Mastodon server admins. I thought there was a discussion about that elsewhere already, but I can't find it now.

Yes! Agreed, we're actively working on this, it's #1305.

**Comment by praichor** [comment]
> But if you sign up for the "LGBTQIA+ Tech Mastodon" instance, you don't expect the kind of visibility on your posts to increase beyond the few thousand people on your local server.

The administrator of this instance can simply block the bridge.

**Comment by snarfed** [comment]
We haven't discussed it publicly much, but apart the policy and product considerations here, another reason we haven't pushed on this more yet is cost. We can self fun Bridgy Fed at its current scale, a bit over 100k users, but if we opened up a big platform like Bluesky (34M+ users!) to be opt-out and automatically on, we worry about the increase in cost.

It wouldn't be immediate, Bridgy Fed would still only do work for people who explicitly interact across the bridge, ie it wouldn't immediately start sending all 34M+ Bluesky users' profiles and posts to fediverse instances. And we don't worry too much about scaling, we think [our current architecture](https://bridgy-fed.readthedocs.io/source/design.html) can hopefully handle at least a 10x increase in load without hitting any structural bottlenecks. But our hosting costs would increase to match, and that's the part we're not quite ready for.

If you're still reading, [we recently launched a Patreon](https://www.patreon.com/c/ANewSocial)! Please consider supporting us there, especially if you already like and use the bridge. If enough people contribute, we may start to feel more ready to pull the trigger here.

**Comment by saschanaz** [comment]
Because that probably doesn't make much sense, that'll cause same bsky accounts to be represented by several different bridgy accounts.

**Comment by lucajet** [comment]
> > why people, maybe some big istance admins, didn't start selfhosting their own Bridgy Fed ?
> 
> How about extending Mastodon to support ATProto in addition to ActivityPub? It wouldn't be the first time Mastodon supports multiple protocols.

Not exactly my point, because of mastodon being very conservative.
But i do i see a space with bridgy fed becoming a service being hosted by admins of mastodon, pleroma, akkoma, pixelfed, whatever.

By analogy, the activitypub plugin that can be installed by every individual running a wordpress blog.

**Comment by osmaa** [comment]
> No, because you don't have to subscribe to more than one bridgy fed servers.

This is a #800 discussion, but identities matter and to a degree they're also an opt-in/opt-out question. While (prospective) followers' desire to see and interact with someone over a bridge is relevant, it would not be correct to duplicate the authors' identity against their wishes, especially if that would lead to confusion about what is the original. That means having multiple BF instances all bridging the same public ATproto accounts for federated visibility to ActivityPub would replicate all the worst parts of the 2022/2023 BirdSiteLive experience, where dozens of same-named accounts, none of which were in control of the original author, propagated through the fediverse.

It might be another thing if a server had their own local-only bridge, but while that might be mostly transparent to a Bluesky native account (who would see follows/replies etc from a number of more-or-less weirdly named bridged AP accounts), it would break federation on ActivityPub - global/public replies would be referring to posts which weren't visible on federated servers, boosts wouldn't federate, and so on.

And of course similar issues would exist in the AP->ATp direction as well, but without even the option of local-only bridging.

It might be desirable to split the BF infrastructure into multiple systems and organizations to reduce a single-point-of-failure risk and delegate costs, but without a solution for the to-be-bridged people to have a choice which of the instances would bridge them, the loss of control over identities would be quite devastating for the integrity of the network.

**Comment by tesaguri** [comment]
>  So far, we are not bridging arbitrary posts, we are bridging opt-in users.

I'm sorry I lost you too, but this very issue is about changing bridges from Bluesky to opt-out basis.

**Comment by saschanaz** [comment]
> I'm very confident that the easiest, quickest, cheapest way to do that is to fund Bridgy Fed itself

What if one of the patreon subscription bonus is to request a given list of bluesky accounts to be federated? $10 for 100 accounts or such?

**Comment by saschanaz** [comment]
And although what @snarfed said is totally valid, it would be nice to solve the technical difficulties of decentralization in parallel, because decentralization is not about cost, it's about power and control. (It would have been much cheaper if we all just stayed in X, but we don't and we have good reasons)

(But let's discuss about this somewhere else, here's not the right place)

**Comment by snarfed** [comment]
> it would be nice to solve the technical difficulties of decentralization in parallel, because decentralization is not about cost, it's about power and control.

Hell yes! Wholeheartedly agree.
